### ISSUE TITLE
Softmax loss function and cross entropy loss function

### ISSUE THREAD

**Issue by segnam** [vlfeat/matconvnet#814]
In vl_nnsoftmaxloss and vl_nnloss, there is a loss function that named softmaxlog loss function

Is it same as cross entropy loss function?

I want to use cross entropy loss function to use in semantic segmentation, but the value of error goes nan

Is there someone who can help me?

**Comment by albanie** [comment]
Short answer: Nearly :).

The `softmax()` part simply normalises your network predictions so that they can be interpreted as probabilities.  Once your network is predicting a probability distribution over labels for each input, the log loss is equivalent to the cross entropy between the true label distribution and the network predictions. The log loss is nicely explained [here](http://www.exegetic.biz/blog/2015/12/making-sense-logarithmic-loss).  

Just as an extra note - `vl_nnsoftmaxloss` is now deprecated - it's functionality has been incorporated into `vl_nnloss`.

**Comment by segnam** [comment]
Thank you for your answering.

So in matconvnet, when i use softmaxlog function, should i give ground truth like below?

for example, three class, 5x5 image
in vl_nnsoftmaxloss(X,c,dzdy)

for an input x
5x5 matrix which passed convolutional neural network

for a ground truth c
5 x 5 pixel labeled matrix 
1 1 1 1 1
1 1 1 1 1
1 2 2 2 1
2 2 2 2 1
3 3 3 3 3 

am i right?

**Comment by albanie** [comment]
Close - the only difference is that you want your network predictions to have the same number of channels there are classes e.g.

        X = rand(5,5,3) ;
        c = [ 1 1 1 1 1 ;
             1 1 1 1 1 ;
             1 2 2 2 1 ;
             2 2 2 2 1 ;
             3 3 3 3 3 ] ;
        y = vl_nnloss(Z, c, [], 'loss', 'softmaxlog') ;

**Comment by segnam** [comment]
Thanks albanie
But I don't know why the network doesn't train
i initialized network network like below
64x64x3 input

5x5x3x32 conv relu 3x3 pool
5x5x32x64 conv relu 3x3 pool
5x5x64x96 conv relu 3x3 pool
5x5x96x128 conv relu 3x3 pool
1x1x128x128 relu
dropout
1x1x128x128 relu
dropout
softmaxloss

after last dropout layer, the output of network is 42x42x6
and i compute error between 42x42 groudtruth which has six class like upper example c
Is there any problem in this training?

**Comment by albanie** [comment]
The outputs of your network need to match the labels.  If you have a fixed size of input that you want to work with, you can modify your network architecture to produce the desired output dimensions.  Alternatively, if you want to fix your network architecture, you can modify the input size to get the desired output size.

As a quick demo, I added a small example to a gist [here](https://gist.github.com/albanie/1030033a96f1b9b10c58d5d4da7ba3c0) which takes the second approach (i.e. uses your architecture but changes the input size to match a segmentation label with spatial dims `5x5`).

**Comment by segnam** [comment]
thanks for your kindness :)

**Comment by segnam** [comment]
Hi i want to detect edge of object in image

And i used modified VGG net to detect edge.
And there is an error while training the network
My network is same as below

input image : 128x128x3
ground truth : 116x116x1
ground truth is labeled as below

1    1   1  1  1 ...
-1 -1  -1-1 -1 ...
-1 -1  -1-1 -1 ...
...
1 means edge and -1 is non-edge

init network
3x3x3x32 relu
3x3x32x32 relu
max pooling

3x3x32x64 relu
3x3x64x64 relu
max pooling

3x3x64x128 relu
3x3x128x128 relu
max pooling

1x1x128x128 relu
1x1x128x1
loss function

logistic loss function is used as loss function

Input image is changed to 116x116, so i gave the ground truth as same size

I don't know why my network is not trained

Filters and values of x is changed to NAN

Please help me

**Comment by idofr** [comment]
@albanie could you please elborate on the normalisation part?
All the function seems to do is to subtract the channel-wise max from the activations and then calculate the log of an hindge loss.

I can neither see the normalisation nor the cross entropy parts
